sbi/simulators/mp_simutils.py

```diff
@@ -0,0 +1,293 @@
+    final_theta = []  # list of parameters
+    minibatches = iterate_minibatches(theta, worker_batch_size)
+
+    pbar = tqdm(total=len(theta), disable=not show_progressbar)
+    desc = "Running {0} simulations".format(len(theta))
+    if type(show_progressbar) == str:
+        desc += show_progressbar
+    pbar.set_description(desc)
+
+    done = False
+    with pbar:
+        while not done:
+            active_list = []
+            for worker, pipe in zip(workers, pipes):
+                try:
+                    theta_worker_batch = next(minibatches)
+                except StopIteration:
+                    # stop passing data to workers
+                    done = True
+                    break
+
+                active_list.append((worker, pipe))
+                log(
+                    "Dispatching to worker (len = {})".format(len(theta_worker_batch)),
+                    verbose,
+                )
+                pipe.send(theta_worker_batch)
+                log("Done", verbose)
+
+            num_remaining = len(active_list)
+            while num_remaining > 0:
+                log("Listening to worker", verbose)
+                msg = queue.get()
+                if isinstance(msg, int):
+                    log("Received int", verbose)
+                elif isinstance(msg, tuple):
+                    log("Received results", verbose)
+                    x, theta = msg
+                    final_x.append(x)
+                    final_theta.append(theta)
+                    num_remaining -= 1
+                    pbar.update(len(theta))
+                else:
+                    log(
+                        "Warning: Received unknown message of type {}".format(
+                            type(msg)
+                        ),
+                        verbose,
+                    )
+
+    stop_workers(workers, pipes, queue, verbose=verbose)
+
+    # n_samples x n_reps x dim theta
+    x = torch.cat(final_x, dim=0)
+    theta = torch.cat(final_theta, dim=0)
+
+    return theta, x
+
+
+def start_workers(simulator: Callable, num_workers: int = 4, verbose: bool = False):
+    """
+    Starting all workers.
+
+    Args:
+        simulator: simulator function that will be executed on each core
+        num_workers: number of parallel workers to start
+        verbose: whether to give information about the current state of the workers
+
+    Returns: queue, workers, pipes
+    """
+    try:
+        start_method = dict(linux="fork", darwin="fork", win32="spawn")
+        mp.set_start_method(start_method[platform])
+    except RuntimeError:
+        # try-except because start method can not be set multiple times within the same
+        # session. So, if the user runs things outside of a __main__() function, e.g. in
+        # a jupyter notebook, this would give an error from the second call of the
+        # set_start_method() function on
+        pass
+
+    pipes = [mp.Pipe(duplex=True) for _ in range(num_workers)]
```

Comment:
I suggest the following to enhance readability:
```python
parents, children = zip(*(mp.Pipe(duplex=True) for _ in range(num_workers)))
worker_cfg = dict(queue=mp.Queue(), simulator=simulator,seed=None, verbose=verbose)
workers = [Worker(i, pipe=child, **worker_cfg) for i, child in enumerate(children)]
(...)
return worker_cfg['queue'], workers, parents
```

Reply:
Had missed this one before, it's in now